lib/ansible/modules/network/dnac/dnac_rest.py

```diff
@@ -0,0 +1,239 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# Copyright: (c) 2018, Kevin Breit (@kbreit) <kevin.breit@kevinbreit.net>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+ANSIBLE_METADATA = {
+    'metadata_version': '1.1',
+    'status': ['preview'],
+    'supported_by': 'community'
+}
+
+DOCUMENTATION = '''
+---
+module: dnac_rest
+
+short_description: Direct access to the Cisco Software-Defined Access (SDA) controller, DNA Central
+
+version_added: "2.6"
+
+description:
+    - Enables management of the Cisco Software-Defined Access controller, more commonly called DNA Center.
+    - More information about the DNA Center API can be found at U(https://developer.cisco.com/site/dna-center-rest-api/).
+
+options:
+    username:
+        description:
+            - The username for use in HTTP basic authentication.
+        required: true
+    password:
+        description:
+            - The password for use in HTTP basic authentication.
+        required: true
+    host:
+        description:
+            - FQDN or IP address of DNA Central server.
+        required: true
+    method:
+        description:
+            - The HTTP method of the request.
+            - Using C(delete) is typically used for deleting objects.
+            - Using C(get) is typically used for querying objects.
+            - Using C(post) is typically used for modifying objects.
+        required: true
+        choices: [ delete, get, post ]
+    path:
+        description:
+            - Directory path to the endpoint. Do not include FQDN specified in C(host).
+        required: true
+    timeout:
+        description:
+            - HTTP timeout value.
+        default: 30
+    use_proxy:
+        description:
+            - If C(no), it will not use a proxy, even if one is defined in an environment variable on the target hosts.
+        type: bool
+        default: 'false'
+    use_ssl:
+        description:
+            - If C(no), it will use HTTP. Otherwise it will use HTTPS.
+        type: bool
+        default: 'true'
+    validate_certs:
+        description:
+            - If C(no), HTTPS certificates will not be verified.
+        type: bool
+        default: 'true'
+    content:
+        description:
+            - Raw content which should be fed in body.
+
+author:
+    - Kevin Breit (@kbreit)
+'''
+
+EXAMPLES = '''
+# Query inventory
+- name: Query network device inventory
+  sda_rest:
+    username: dnacuser
+    password: dnaccpass
+    host: dnac.yourorg.com
+    method: get
+    path: '/api/v1/network-device'
+    use_ssl: Yes
+  delegate_to: localhost
+
+'''
+
+RETURN = '''
+message:
```

Comment:
I think `message` is a strange name for the resulting output
Do you have example output ?

Reply:
Here is a snippet of the output. The rest is pretty similar.

`msg": {
        "changed": true,
        "failed": false,
        "message": {
            "response": [
                {
                    "apManagerInterfaceIp": "",
                    "associatedWlcIp": "",
                    "bootDateTime": "2018-01-11 15:47:04",
                    "collectionInterval": "Global Default",
                    "collectionStatus": "Managed",
                    "errorCode": null,
                    "errorDescription": null,
                    "family": "Routers",
                    "hostname": "asr1001-x.abc.inc",
                    "id": "d5bbb4a9-a14d-4347-9546-89286e9f30d4",
                    "instanceUuid": "d5bbb4a9-a14d-4347-9546-89286e9f30d4",
                    "interfaceCount": "10",
                    "inventoryStatusDetail": "<status><general code=\"SUCCESS\"/></status>",
                    "lastUpdateTime": 1519355055503,
                    "lastUpdated": "2018-02-23 03:04:15",
                    "lineCardCount": "9",`

Reply:
So what I would do is merge the output with the result, so that in case you do:

```yaml
    register: this
```

You can access the output in `this.response.bootDateTime`. There's no added value to have to use `this.message.response.bootDateTime` and this will make it more convenient to users.

Are there any other root-keys in the output, apart from `response` ?

Reply:
It would be nice to have the complete output (maybe in gist) to understand what the data structure looks like.

Reply:
Admittedly, I'm not entirely clear on what value I should be putting in the documentation for this. But here is a response from DNA Center with a full output. In this case it is for inventory. Of course, since this is a generic REST module, other outputs may look different.

https://gist.github.com/kbreit/87b3bc4d9cbb5564d493045f695c23aa

Reply:
Ok, so I see 2 values being returned, `response` and `version`. I would merged the result dictionary and add both.

```python
    result.update(json.loads(resp.read()))
```

Reply:
The `version` return value could be valuable for users that need to support more than one DNA-Center which may return different results.

Reply:
I'm updating the documentation to assume it's the API version. The DNA Controller version is 1.1.2 so I can't tell exactly what the version is referring to. I'll need to see which version of DNA Controller that API version 1.0 associates to.

Reply:
It's the version of the returned data. As long as they don't make any incompatible update to the output, it will stay at 1.0. It does not need to map to a specific API version though. Although if at some point they change the output, it probably will be related to a specific new (major) release.